Do not attempt to set breakpoints on files extensions other than .ps1 or .psm1

Got this while trying to debug a PS script in the PowerShell repo where I had previously been debugging some C# code.  So I had a mix of .cs and .ps1 breakpoints.  I noticed this in my log file:
```
11/4/2017 1:10:59 PM [VERBOSE] - Method "ExecuteCommand" at line 491 of C:\projects\powershelleditorservices\src\PowerShellEditorServices\Session\PowerShellContext.cs

    Attempting to execute command(s):
    
        Microsoft.PowerShell.Utility\Set-PSBreakpoint
    

11/4/2017 1:10:59 PM [VERBOSE] - Method "OnSessionStateChanged" at line 1138 of C:\projects\powershelleditorservices\src\PowerShellEditorServices\Session\PowerShellContext.cs

    Session state changed --
    
        Old state: Running
        New state: Ready
        Result: Completed

11/4/2017 1:10:59 PM [ERROR] - Method "ExecuteCommand" at line 545 of C:\projects\powershelleditorservices\src\PowerShellEditorServices\Session\PowerShellContext.cs

    Execution completed with errors:
    
    A parameter cannot be found that matches parameter name 'StringBuilder'.
    A parameter cannot be found that matches parameter name 'StringBuilder'.
    A parameter cannot be found that matches parameter name 'StringBuilder'.
    A parameter cannot be found that matches parameter name 'StringBuilder'.
    Cannot set breakpoint on file 'C:\Users\Keith\GitHub\rkeithhill\PowerShell\src\System.Management.Automation\engine\parser\tokenizer.cs'; only *.ps1 and *.psm1 files are valid.
    A parameter cannot be found that matches parameter name 'StringBuilder'.
    

11/4/2017 1:10:59 PM [VERBOSE] - Method "OnSessionStateChanged" at line 1138 of C:\projects\powershelleditorservices\src\PowerShellEditorServices\Session\PowerShellContext.cs
```
Seems pretty obvious that we shouldn't bother to set PowerShell breakpoints on these other file types.

--- a/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs
+++ b/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs
@@ -510,9 +510,34 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
                         Breakpoints = srcBreakpoints.ToArray()
                     });
 
-	            return;
-	        }
+                return;
+            }
 
+            // Verify source file is a PowerShell script file.
+            string fileExtension = Path.GetExtension(scriptFile?.FilePath ?? "")?.ToLower();
+            if (string.IsNullOrEmpty(fileExtension) || ((fileExtension != ".ps1") && (fileExtension != ".psm1")))
+            {
+                Logger.Write(
+                    LogLevel.Warning,
+                    $"Attempted to set breakpoints on a non-PoweShell file: {setBreakpointsParams.Source.Path}");
+
+                string message = this.noDebug ? string.Empty : "Source is not a PowerShell script, breakpoint not set.";
+
+                var srcBreakpoints = setBreakpointsParams.Breakpoints
+                    .Select(srcBkpt => Protocol.DebugAdapter.Breakpoint.Create(
+                        srcBkpt, setBreakpointsParams.Source.Path, message, verified: this.noDebug));
+
+                // Return non-verified breakpoint message.
+                await requestContext.SendResult(
+                    new SetBreakpointsResponseBody
+                    {
+                        Breakpoints = srcBreakpoints.ToArray()
+                    });
+
+                return;
+            }
+
+            // At this point, the source file has been verified as a PowerShell script.
             var breakpointDetails = new BreakpointDetails[setBreakpointsParams.Breakpoints.Length];
             for (int i = 0; i < breakpointDetails.Length; i++)
             {
@@ -541,7 +566,7 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
                 catch (Exception e)
                 {
                     // Log whatever the error is
-                    Logger.WriteException("Caught error while setting breakpoints in SetBreakpoints handler", e);
+                    Logger.WriteException($"Caught error while setting breakpoints in SetBreakpoints handler for file {scriptFile?.FilePath}", e);
                 }
                 finally
                 {

--- a/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs
+++ b/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs
@@ -87,7 +87,7 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
             // Is this an untitled script?
             Task launchTask = null;
 
-            if (this.scriptToLaunch.StartsWith("untitled:"))
+            if (ScriptFile.IsUntitledPath(this.scriptToLaunch))
             {
                 ScriptFile untitledScript =
                     this.editorSession.Workspace.GetFile(
@@ -259,7 +259,7 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
 
             // When debugging an "untitled" (unsaved) file - the working dir can't be derived
             // from the Script path.  OTOH, if the launch params specifies a Cwd, use it.
-            if (workingDir.StartsWith("untitled:") && string.IsNullOrEmpty(launchParams.Cwd))
+            if (ScriptFile.IsUntitledPath(workingDir) && string.IsNullOrEmpty(launchParams.Cwd))
             {
                 workingDir = null;
             }
@@ -492,7 +492,7 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
             {
                 // When you set a breakpoint in the right pane of a Git diff window on a PS1 file,
                 // the Source.Path comes through as Untitled-X.
-                if (!setBreakpointsParams.Source.Path.StartsWith("untitled:"))
+                if (!ScriptFile.IsUntitledPath(setBreakpointsParams.Source.Path))
                 {
                     scriptFile = editorSession.Workspace.GetFile(setBreakpointsParams.Source.Path);
                 }

--- a/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs
+++ b/src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs
@@ -264,7 +264,7 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
                 workingDir = null;
             }
 
-            if (workingDir != null)
+            if (!string.IsNullOrEmpty(workingDir))
             {
                 workingDir = PowerShellContext.UnescapePath(workingDir);
                 try
@@ -282,7 +282,7 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
                 }
             }
 
-            if (workingDir == null)
+            if (string.IsNullOrEmpty(workingDir))
             {
 #if CoreCLR
                 workingDir = AppContext.BaseDirectory;
@@ -490,7 +490,12 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
             // VSCode sends breakpoint requests with the original filename that doesn't exist anymore.
             try
             {
-                scriptFile = editorSession.Workspace.GetFile(setBreakpointsParams.Source.Path);
+                // When you set a breakpoint in the right pane of a Git diff window on a PS1 file,
+                // the Source.Path comes through as Untitled-X.
+                if (!setBreakpointsParams.Source.Path.StartsWith("untitled:"))
+                {
+                    scriptFile = editorSession.Workspace.GetFile(setBreakpointsParams.Source.Path);
+                }
             }
             catch (Exception e) when (e is FileNotFoundException || e is DirectoryNotFoundException)
             {
@@ -519,7 +524,7 @@ namespace Microsoft.PowerShell.EditorServices.Protocol.Server
             {
                 Logger.Write(
                     LogLevel.Warning,
-                    $"Attempted to set breakpoints on a non-PoweShell file: {setBreakpointsParams.Source.Path}");
+                    $"Attempted to set breakpoints on a non-PowerShell file: {setBreakpointsParams.Source.Path}");
 
                 string message = this.noDebug ? string.Empty : "Source is not a PowerShell script, breakpoint not set.";
 

--- a/src/PowerShellEditorServices/Workspace/ScriptFile.cs
+++ b/src/PowerShellEditorServices/Workspace/ScriptFile.cs
@@ -208,6 +208,19 @@ namespace Microsoft.PowerShell.EditorServices
         }
 
         /// <summary>
+        /// Deterines whether the supplied path indicates the file is an "untitled:Unitled-X" 
+        /// which has not been saved to file.
+        /// </summary>
+        /// <param name="path">The path to check.</param>
+        /// <returns>True if the path is an untitled file, false otherwise.</returns>
+        public static bool IsUntitledPath(string path)
+        {
+            Validate.IsNotNull(nameof(path), path);
+
+            return path.ToLower().StartsWith("untitled:");
+        }
+
+        /// <summary>
         /// Gets a line from the file's contents.
         /// </summary>
         /// <param name="lineNumber">The 1-based line number in the file.</param>

